File: src/prod/src/data/txnreplicator/statemanager/SerializableMetadata.cpp
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

#include "stdafx.h"

using namespace Data;
using namespace Data::StateManager;

NTSTATUS SerializableMetadata::Create(
    __in KUri const & name,
    __in KString const & typeString,
    __in FABRIC_STATE_PROVIDER_ID stateProviderId,
    __in FABRIC_STATE_PROVIDER_ID parentStateProviderId,
    __in FABRIC_SEQUENCE_NUMBER createLSN,
    __in FABRIC_SEQUENCE_NUMBER deleteLSN,
    __in MetadataMode::Enum metadataMode,
    __in KAllocator & allocator,
    __in_opt Utilities::OperationData const * const initializationParameters,
    __out CSPtr & result) noexcept
{
    // EmptyStateProviderId is state manager id, thus state provider id can't be EmptyStateProviderId
    ASSERT_IFNOT(stateProviderId != Constants::EmptyStateProviderId, "Empty state provider id specified");

    result = _new(SERIALIZABLEMETADATA_TAG, allocator) SerializableMetadata(name, typeString, stateProviderId, parentStateProviderId, createLSN, deleteLSN, metadataMode, initializationParameters);

    if (result == nullptr)
    {
        return STATUS_INSUFFICIENT_RESOURCES;
    }

    if (!NT_SUCCESS(result->Status()))
    {
        // null Result while fetching failure status with no extra AddRefs or Releases; should opt very well
        return (CSPtr(Ktl::Move(result)))->Status();
    }

    return STATUS_SUCCESS;
}

NTSTATUS SerializableMetadata::Create(
    __in Metadata const & metadata, 
    __in KAllocator & allocator,
    __out CSPtr & result) noexcept
{
    return Create(
        *metadata.Name,
        *metadata.TypeString,
        metadata.StateProviderId,
        metadata.ParentId,
        metadata.CreateLSN,
        metadata.DeleteLSN,
        metadata.MetadataMode,
        allocator,
        metadata.InitializationParameters.RawPtr(),
        result);
}

// Property
KUri::CSPtr SerializableMetadata::get_Name() const { return name_; }

KString::CSPtr SerializableMetadata::get_TypeString() const { return typeString_; }

Utilities::OperationData::CSPtr SerializableMetadata::get_InitializationParameters() const { return initializationParameters_; }

FABRIC_STATE_PROVIDER_ID SerializableMetadata::get_StateProviderId() const { return stateProviderId_; }

FABRIC_STATE_PROVIDER_ID SerializableMetadata::get_ParentStateProviderId() const { return parentStateProviderId_; }

FABRIC_SEQUENCE_NUMBER SerializableMetadata::get_CreateLSN() const { return createLSN_; }

FABRIC_SEQUENCE_NUMBER SerializableMetadata::get_DeleteLSN() const { return deleteLSN_; }

MetadataMode::Enum SerializableMetadata::get_MetadataMode() const { return metadataMode_; }

SerializableMetadata::SerializableMetadata(
    __in KUri const & name,
    __in KString const & typeString,
    __in FABRIC_STATE_PROVIDER_ID stateProviderId,
    __in FABRIC_STATE_PROVIDER_ID parentStateProviderId,
    __in FABRIC_SEQUENCE_NUMBER createLSN,
    __in FABRIC_SEQUENCE_NUMBER deleteLSN,
    __in MetadataMode::Enum metadataMode,
    __in_opt Utilities::OperationData const * const initializationParameters) noexcept
    : KObject()
    , KShared()
    , name_(&name)
    , typeString_(&typeString)
    , stateProviderId_(stateProviderId)
    , parentStateProviderId_(parentStateProviderId)
    , createLSN_(createLSN)
    , deleteLSN_(deleteLSN)
    , metadataMode_(metadataMode) 
    , initializationParameters_(initializationParameters)
{
}

SerializableMetadata::~SerializableMetadata()
{
}

File: src/prod/src/managed/Microsoft.ServiceFabric.Data.Impl/Replicator/SerializableMetadata.cs
// ------------------------------------------------------------
// Copyright (c) Microsoft Corporation.  All rights reserved.
// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
// ------------------------------------------------------------

namespace Microsoft.ServiceFabric.Replicator
{
    using System;

    /// <summary>
    /// Metadata for state manager local state
    /// </summary>
    internal class SerializableMetadata
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="SerializableMetadata"/> class. 
        /// </summary>
        public SerializableMetadata(
            Uri name,
            Type type,
            byte[] initializationParameters,
            long stateProviderId,
            long parentStateProviderId,
            MetadataMode metadataMode,
            long createLsn,
            long deleteLsn) : this(name, initializationParameters, stateProviderId, parentStateProviderId, metadataMode, createLsn, deleteLsn)
        {
            Utility.Assert(type != null, "type cannot be null.");
            this.Type = type;
            this.TypeString = null;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="SerializableMetadata"/> class. 
        /// </summary>
        public SerializableMetadata(
            Uri name,
            string type,
            byte[] initializationParameters,
            long stateProviderId,
            long parentStateProviderId,
            MetadataMode metadataMode,
            long createLsn,
            long deleteLsn) : this(name, initializationParameters, stateProviderId, parentStateProviderId, metadataMode, createLsn, deleteLsn)
        {
            Utility.Assert(type != null, "type cannot be null.");
            this.TypeString = type;
            this.Type = null;
        }

        /// <summary>
        /// Initializes a new instance of the <see cref="SerializableMetadata"/> class. 
        /// </summary>
        internal SerializableMetadata(
            Uri name,
            byte[] initializationParameters,
            long stateProviderId,
            long parentStateProviderId,
            MetadataMode metadataMode,
            long createLsn,
            long deleteLsn)
        {
            Utility.Assert(name != null, "Name cannot be empty in replication metadata.");

            Utility.Assert(stateProviderId != DynamicStateManager.EmptyStateProviderId, "state provider id cannot be empty.");

            this.Name = name;

            this.CreateLsn = createLsn;
            this.DeleteLsn = deleteLsn;

            this.InitializationContext = initializationParameters;
            this.StateProviderId = stateProviderId;
            this.ParentStateProviderId = parentStateProviderId;
            this.MetadataMode = metadataMode;
        }

        /// <summary>
        /// Gets or sets the create sequence number.
        /// </summary>
        internal long CreateLsn { get; private set; }

        /// <summary>
        /// Gets or sets the delete sequence number.
        /// </summary>
        internal long DeleteLsn { get; set; }

        /// <summary>
        /// Gets or sets the initialization parameters.
        /// </summary>
        internal byte[] InitializationContext { get; private set; }

        /// <summary>
        /// Gets or sets the metadata mode.
        /// </summary>
        internal MetadataMode MetadataMode { get; set; }

        /// <summary>
        /// Gets or sets the name.
        /// </summary>
        internal Uri Name { get; private set; }

        /// <summary>
        /// Gets or sets the id for this state provider's parent (or Guid.Empty if it has no parent).
        /// </summary>
        internal long ParentStateProviderId { get; private set; }

        /// <summary>
        /// Gets or sets the state provider id.
        /// </summary>
        internal long StateProviderId { get; private set; }

        /// <summary>
        /// Gets or sets the type.
        /// </summary>
        internal Type Type { get; private set; }

        /// <summary>
        /// Gets or sets the type string.
        /// </summary>
        internal string TypeString { get; private set; }
    }
}